lib/parser/provenance.cc

```diff
@@ -14,10 +14,51 @@
 
 #include "provenance.h"
 #include "../common/idioms.h"
+#include <algorithm>
 #include <utility>
 
 namespace Fortran::parser {
 
+ProvenanceRangeToOffsetMappings::ProvenanceRangeToOffsetMappings() {}
+ProvenanceRangeToOffsetMappings::~ProvenanceRangeToOffsetMappings() {}
+
+void ProvenanceRangeToOffsetMappings::Put(
+    ProvenanceRange range, std::size_t offset) {
+  auto fromTo{map_.equal_range(range)};
+  for (auto iter{fromTo.first}; iter != fromTo.second; ++iter) {
+    if (range == iter->first) {
+      iter->second = std::min(offset, iter->second);
+      return;
+    }
+  }
+  if (fromTo.second != map_.end()) {
+    map_.emplace_hint(fromTo.second, range, offset);
+  } else {
+    map_.emplace(range, offset);
+  }
+}
+
+std::optional<std::size_t> ProvenanceRangeToOffsetMappings::Map(
+    ProvenanceRange range) const {
```

Comment:
Since the function only maps the start of the `ProvenanceRange` to offset, maybe the input should only be `Provenance`? This would make it easier if the caller only cares about finding the offset of the start of the `ProvenanceRange`. And it avoids the situation where the caller accidentally uses a larger than necessary `ProvenanceRange`.

Reply:
The `multimap` was configured with a comparator that will cause the call to `equal_range(range)` to return a range of iterators that might overlap the entirety of `range`, not just its first position.  This was intentional.  In the event that a client wants to map only a single byte, a singleton range can be used.

Reply:
My understanding is that if the caller calls this function with a range that doesn't fit any `ProvenanceRange` inside the map, even if the start of the input range is inside of the ProvenanceRange.
Is there a situation where the caller would prefer to call this function with a full `ProvenanceRange` over calling it with a singleton range? Calling this function with the full range has the pitfall of giving it too large a range and whiffing. I think we should ask for a single Provenance and convert it to a ProvenanceRange for the caller.

Reply:
Passing a range of multiple byte positions assures the caller that any returned offset points to a corresponding contiguous range in the cooked character stream -- `FOO` in the editor gets mapped to a `foo` in the CCS.  Passing a single position (`F`) would return a result that just points to the `f`, and that's not what one would be probing the symbol table with later.

(What this code can't do is map fixed-form `F O O` to `foo`, and maybe it should.  Same with identifiers broken across line continuations.)

Reply:
Ah, I understand what you're saying. The caller would be assured that the range passed in corresponds to the range they expect if it hits, even if the function only returns the start of it. Thank you for the explanations.